lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,140 @@
+require 'open3'
+require 'thread'
+
+module PuppetX
+  module Dsc
+    class PowerShellManager
+      attr_accessor :return_code
+
+      @@cmd_instances = {}
+
+      def self.instance(cmd)
+        @@cmd_instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        process_init_lock = Mutex.new
+        process_started = ConditionVariable.new
+
+        @stdout_lock = Mutex.new
+        @stdout_written = ConditionVariable.new
+
+        @stdout_capture = []
+        stdout_monitor_thread = nil
+
+        process_init_lock.synchronize do
+          @runner_thread = Thread.new do
+            Open3::popen2(cmd) do |stdin, stdout, proc_thread|
+              puts "#{Time.now} #{cmd} is running as pid: #{proc_thread.pid}"
+
+              @stdin = stdin
+              @stdout = stdout
+              stdin.sync = true
+              stdout.sync = true
+
+              stdout_monitor_thread = Thread.new do
+                while !@stdout.eof? do
+                  puts "monitor thread - entering stdout_lock synchronize"
+                  @stdout_lock.synchronize do
+                    begin
+                      # this is necessary to ensure all output captured
+                      @stdout.flush
+                      while self.class.is_readable?(@stdout) do
+                        if !proc_thread.alive?
+                          puts "process has exited so stdout is bad"
+                          raise "Process has exited, no stdout to give"
+                        end
```

Comment:
Of course this isn't quite the right message yet, still rough as I just determined what the issue was.